tests/functional/test_package.py

```diff
@@ -23,6 +31,257 @@ def index():
+        reqs = ['foo', 'bar', 'baz']
+        pip, runner = pip_runner
+        pip.add_side_effect('download', [
+            PipSideEffect('foo-1.0-cp36-none-any.whl'),
+            PipSideEffect('bar-1.2-cp36-cp36m-manylinux1_x86_64.whl'),
+            PipSideEffect('baz-1.5-cp36-cp36m-linux_x86_64.whl')
+        ])
+
+        appdir = str(_create_app_structure(tmpdir))
+        self._write_requirements_txt('\n'.join(reqs), appdir)
+        builder = DependencyBuilder(osutils, runner)
+        site_packages = builder.build_site_packages(appdir)
+        installed_packages = os.listdir(site_packages)
+        for req in reqs:
+            assert req in installed_packages
+
+    def test_can_get_py27_whls(self, tmpdir, osutils, pip_runner):
+        reqs = ['foo', 'bar', 'baz']
+        pip, runner = pip_runner
+        pip.add_side_effect('download', [
+            PipSideEffect('foo-1.0-cp27-none-any.whl'),
+            PipSideEffect(
+                'bar-1.2-cp27-none-manylinux1_x86_64.whl'),
+            PipSideEffect('baz-1.5-cp27-cp27mu-linux_x86_64.whl')
+        ])
+
+        appdir = str(_create_app_structure(tmpdir))
+        self._write_requirements_txt('\n'.join(reqs), appdir)
+        builder = DependencyBuilder(osutils, runner)
+        site_packages = builder.build_site_packages(appdir)
+        installed_packages = os.listdir(site_packages)
+        for req in reqs:
+            assert req in installed_packages
+
+    def test_does_fail_on_narrow_py27_unicode(self, tmpdir, osutils,
+                                              pip_runner):
+        reqs = ['baz']
+        pip, runner = pip_runner
+        pip.add_side_effect('download', [
+            PipSideEffect('baz-1.5-cp27-cp27m-linux_x86_64.whl')
+        ])
+
+        appdir = str(_create_app_structure(tmpdir))
+        self._write_requirements_txt('\n'.join(reqs), appdir)
+        builder = DependencyBuilder(osutils, runner)
+        with consume_stdout_and_stderr() as (out, _):
+            site_packages = builder.build_site_packages(appdir)
+        installed_packages = os.listdir(site_packages)
+        assert len(installed_packages) == 0
+        assert 'Could not install dependencies:\nbaz==1.5' in out.getvalue()
+
+    def test_does_fail_on_python_1_whl(self, tmpdir, osutils, pip_runner):
+        reqs = ['baz']
+        pip, runner = pip_runner
+        pip.add_side_effect('download', [
+            PipSideEffect('baz-1.5-cp14-cp14m-linux_x86_64.whl')
+        ])
+
+        appdir = str(_create_app_structure(tmpdir))
+        self._write_requirements_txt('\n'.join(reqs), appdir)
+
+        builder = DependencyBuilder(osutils, runner)
+        with consume_stdout_and_stderr() as (out, _):
+            site_packages = builder.build_site_packages(appdir)
+        installed_packages = os.listdir(site_packages)
+        assert len(installed_packages) == 0
+        assert 'Could not install dependencies:\nbaz==1.5' in out.getvalue()
```

Comment:
Thinking about this again. I wonder if there is a better way to represent the packages that failed to install than relying on standard error? I wonder if we should be raising an error or returning the list of failed packages (along with successfully installed packages)?